src/stk500.c

```diff
@@ -557,6 +569,73 @@ static int stk500_initialize(const PROGRAMMER *pgm, const AVRPART *p) {
     }
   }
 
+  // Read or write target voltage
+  if (PDATA(pgm)->vtarg_get || PDATA(pgm)->vtarg_set) {
+    // Read current target voltage set value
+    unsigned int vtarg_read;
+    stk500_getparm(pgm, Parm_STK_VTARGET, &vtarg_read);
+    if (PDATA(pgm)->vtarg_get)
+      msg_info("Target voltage value read as %.2fV\n", (vtarg_read / 10.0));
+    // Write target voltage value
+    else {
+      msg_info("Changing target voltage from %.2f to %.2fV\n", (vtarg_read / 10.0), PDATA(pgm)->vtarg_data);
+      if(pgm->set_vtarget(pgm, PDATA(pgm)->vtarg_data) < 0)
+        return -1;
+    }
+  }
+
+  // Read or write analog reference voltage
+  if (PDATA(pgm)->varef_get || PDATA(pgm)->varef_set) {
+    // Read current analog reference voltage
+    unsigned int varef_read;
+    stk500_getparm(pgm, Parm_STK_VADJUST, &varef_read);
+    if (PDATA(pgm)->varef_get)
+      msg_info("Analog reference voltage value read as %.2fV\n", (varef_read / 10.0));
+    // Write analog reference voltage
+    else {
+      msg_info("Changing analog reference voltage from %.2f to %.2fV\n",
+        (varef_read / 10.0), PDATA(pgm)->varef_data);
+      if(pgm->set_varef(pgm, 0, PDATA(pgm)->varef_data) < 0)
+        return -1;
+    }
+  }
+
+  // Read or write clock generator frequency
+  if (PDATA(pgm)->fosc_get || PDATA(pgm)->fosc_set) {
+    // Read current target voltage set value
+    unsigned int osc_pscale;
+    unsigned int osc_cmatch;
+    const char *unit_get = {"Hz"};
+    double f_get;
+    stk500_getparm(pgm, Parm_STK_OSC_PSCALE, &osc_pscale);
+    stk500_getparm(pgm, Parm_STK_OSC_CMATCH, &osc_cmatch);
+    if(osc_pscale) {
+      int prescale = 1;
+      f_get = STK500_XTAL / 2;
+      switch (osc_pscale) {
+        case 2: prescale = 8; break;
+        case 3: prescale = 32; break;
+        case 4: prescale = 64; break;
+        case 5: prescale = 128; break;
+        case 6: prescale = 256; break;
+        case 7: prescale = 1024; break;
+      }
+      f_get /= prescale;
+      f_get /= (osc_cmatch + 1);
+      f_get = f_to_kHz_MHz(f_get, &unit_get);
+    }
+    if (PDATA(pgm)->fosc_get)
```

Comment:
Might f_get end up being used uninitialised here? In case osc_pscale is false and `PDATA(pgm)->fosc_get` is there?

Reply:
Absolutely, good catch!
It prints "0.000 Hz" even though it's not bein initialized, but it's a good rule of thumb anyways.